hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMapProxy.java

```diff
@@ -1267,12 +1270,30 @@ public Object executeOnKey(K key, EntryProcessor entryProcessor) {
     }
 
     public Object executeOnKeyInternal(Data keyData, EntryProcessor entryProcessor) {
+        validateEntryProcessorForSingleKeyProcessing(entryProcessor);
         ClientMessage request = MapExecuteOnKeyCodec.encodeRequest(name, toData(entryProcessor), keyData, getThreadId());
         ClientMessage response = invoke(request, keyData);
         MapExecuteOnKeyCodec.ResponseParameters resultParameters = MapExecuteOnKeyCodec.decodeResponse(response);
         return toObject(resultParameters.response);
     }
 
+    private static void validateEntryProcessorForSingleKeyProcessing(EntryProcessor entryProcessor) {
+        if (entryProcessor instanceof Offloadable) {
+            EntryBackupProcessor backupProcessor = entryProcessor.getBackupProcessor();
+            if (!(backupProcessor instanceof Offloadable)) {
```

Comment:
Isn't `if (!(backupProcessor instanceof Offloadable))` conflicting with the `ReadOnly` check below? When EP is `ReadOnly`, you expect  `getBackupProcessor()` to return `null` but  `(null instanceof Offloadable)` returns `false`. 

It looks like you should swap the places of `if (entryProcessor instanceof Offloadable) {...` and `if (entryProcessor instanceof ReadOnly) {...` blocks.